### PR TITLE
feat(versioning): Add support for Ubuntu version scheme

### LINKED ISSUE #6060: Support ubuntu versioning
**What would you like Renovate to be able to do?**

Know that `.04` releases are stable and `.10` are unstable.

**Describe the solution you'd like**

Maybe a simple versioning wrapper called `ubuntu`.

**Describe alternatives you've considered**

Users need to manually close PRs to ignore them.

### PR DESCRIPTION
**Pull request by zharinov**
Closes #6060

### CODE REVIEW

**Comment by zharinov** [comment]
For now, it's focused around fixed set of Ubuntu releases.
I'm not sure to which degree we need to make assumptions about future versions.
Maybe we may query and parse [this endpoint](https://changelogs.ubuntu.com/meta-release) at Renovate build-time.

File: lib/versioning/ubuntu/index.spec.ts [hunk 1]
```diff
@@ -0,0 +1,448 @@
...
+    expect(ubuntu.isStable('16.04.7')).toBe(true); // LTS
+    expect(ubuntu.isStable('16.10')).toBe(false);
+    expect(ubuntu.isStable('17.04')).toBe(false);
+    expect(ubuntu.isStable('17.10')).toBe(false);
+    expect(ubuntu.isStable('18.04.5')).toBe(true); // LTS
+    expect(ubuntu.isStable('18.10')).toBe(false);
+    expect(ubuntu.isStable('19.04')).toBe(false);
+    expect(ubuntu.isStable('19.10')).toBe(false);
+    expect(ubuntu.isStable('20.04')).toBe(true); // LTS
+    expect(ubuntu.isStable('2020.04')).toBe(false);
```
Comment (rarkins) [inline lib/versioning/ubuntu/index.spec.ts:166]: Please test the next expected unstable one

File: lib/versioning/ubuntu/index.spec.ts [hunk 2]
```diff
@@ -0,0 +1,448 @@
+import { api as ubuntu } from '.';
+
+describe('versioning/ubuntu', () => {
+  // validation
+
+  it('isCompatible', () => {
+    expect(ubuntu.isCompatible(undefined)).toBe(false);
```
Comment (rarkins) [inline lib/versioning/ubuntu/index.spec.ts:7]: I need to think how we're using isCompatible and if this is the right way

**Review by rarkins — CHANGES_REQUESTED** [review]

**Comment by viceice** [comment]
Ubuntu has some conventions. So every even year on April we have a lts release. I've not found any exemption until now

**Comment by zharinov** [comment]
The only exception I see is older `6.06 Dapper Drake`, which I could just hardcode. If there is potential need for fancy names decoding, we should periodically update them. If not, then we can compute all future versions according to conventions.

**Comment by rarkins** [comment]
I'm not interested in the code name upgrading now. Just that we recognise unstable

**Comment by viceice** [comment]
I think we don't need those old versions, as nobody should use them anymore

**Comment by zharinov** [comment]
Also, we have irregular patch versions like `18.04.5`, `16.04.7`, `14.04.6`, etc. Are we okay to keep it simple and consider any ubuntu-like versions, including not actually released like `18.04.1`, `18.04.2`, `18.04.3`, ...?

**Comment by viceice** [comment]
I think we should move the versions declaration to a Ubuntu datasource, so we can hopefully fetch them on the fly

**Comment by viceice** [comment]
Sure, we want to update to latest patch versions later 🙃

**Comment by rarkins** [comment]
Yes make the patch part dynamic

**Comment by zharinov** [comment]
Test repo: https://github.com/zharinov/renovate-test-ubuntu/pulls

File: lib/versioning/ubuntu/index.ts [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+function isVersion(input: string): string | boolean | null {
+  return isValid(input);
+}
+
+function isCompatible(
+  version: string,
+  _range?: string
+): string | boolean | null {
+  return isValid(version);
```
Comment (zharinov) [inline lib/versioning/ubuntu/index.ts:25]: Not sure what option is the best here

**Review by viceice — COMMENTED** [review]
i think this is ok now. But why PR zharinov/renovate-test-ubuntu#3 will upgrate to 20.10 instead of stable 20.04?

**Comment by zharinov** [comment]
> i think this is ok now. But why PR zharinov/renovate-test-ubuntu#3 will upgrate to 20.10 instead of stable 20.04?

@viceice Sorry, can you remember from which version it was upgrading? The cleaning of my forks went too far 😅

**Comment by viceice** [comment]
I think it was 18.04

**Comment by rarkins** [comment]
The only scenario where we should upgrade to `20.10` would be if the user configures `ignoreUnstable=false`.

**Comment by zharinov** [comment]
Well, now it doesn't upgrade at all. I haven't yet figured out why.

Current implementation:
```json
DEBUG: packageFiles with updates (repository=zharinov/renovate-test-ubuntu)
                 ...
                 "updates": [
                   {
                     "fromVersion": "16.04",
                     "toVersion": "16.10",
                     "newValue": "16.10",
                     "newMajor": 16,
                     "newMinor": 10,
                     "updateType": "minor",
                     "isSingleVersion": true,
                     "newVersion": "16.10"
                   },
                   {
                     "fromVersion": "16.04",
                     "toVersion": "21.04",
                     "newValue": "21.04",
                     "newMajor": 21,
                     "newMinor": 4,
                     "updateType": "major",
                     "isSingleVersion": true,
                     "newVersion": "21.04"
                   }
                 ],
                 ...
```

This implementation:
```json
DEBUG: packageFiles with updates (repository=zharinov/renovate-test-ubuntu)
                 ...
                 "updates": [
                   {
                     "fromVersion": "16.04",
                     "toVersion": "20.04",
                     "newValue": "20.04",
                     "newMajor": 20,
                     "newMinor": 4,
                     "updateType": "major",
                     "isSingleVersion": true
                   }
                 ],
                 ...
```

**Comment by viceice** [comment]
Maybe we need some schedules (like node) to define a version as stable? I'm pretty sure `21.04` is not yet released.

**Comment by rarkins** [comment]
> Maybe we need some schedules (like node) to define a version as stable? I'm pretty sure `21.04` is not yet released.

But if it's not released then would any datasource return it in the first place?

**Comment by rarkins** [comment]
Whoops, here they are: https://hub.docker.com/_/ubuntu?tab=tags&page=1&ordering=last_updated

**Comment by rarkins** [comment]
So yes, we need to have a schedule file for each stable pattern and (for now) keep it updated manually.

**Comment by viceice** [comment]
Maybe we can simply mark as unstable if not released? so if users want to have updated to this, he can configure `ignoreUnstable=false` and will get 21.04

**Comment by rarkins** [comment]
Yes, I think we can make this simple:

- Anything version matching `^\d+\.04` defaults to stable, otherwise unstable
- If starting with `21.04` then it's unstable if date < April 22, 2021

Let's ignore "named" versions for now.

Anything wrong with that which I'm missing?

**Comment by viceice** [comment]
> Yes, I think we can make this simple:
> 
> * Anything version matching `^\d+\.04` defaults to stable, otherwise unstable
> * If starting with `21.04` then it's unstable if date < April 22, 2021
> 
> Let's ignore "named" versions for now.
> 
> Anything wrong with that which I'm missing?

19.04 isn't stable, only even years are lts, so 16.04, 18.04, 20.04

**Comment by rarkins** [comment]
whoops, I forgot about odds ones. Simpler revised rule:

- If matches `/^\d?(0|2|4|6|8)\\.04/` then stable

And sometime in the future we can add a custom rule for 22.04 based on schedule

**Comment by viceice** [comment]
if we use schedules, we should also mark unsupported lts versions as unstable

**Comment by rarkins** [comment]
> if we use schedules, we should also mark unsupported lts versions as unstable

That would better fit in as *deprecated*, not unstable. Currently deprecated details is sourced from the datasource only though

**Comment by viceice** [comment]
sure, i used wrong wording 🤦‍♂️

**Comment by viceice** [comment]
> Yes, fixing Ubuntu and Node versioning are our pre-requisites for removing that restriction :)

java has a bad versioning too, eg v8 and v11 are lts, next lts will be v18.

https://en.wikipedia.org/wiki/Java_version_history

**Comment by rarkins** [comment]
That one seems easier..

**Comment by zharinov** [comment]
So, can we implement schedule-based things in separate PR?
I would like to finish _static_ part of implementation (though, maybe not use it for now in `docker/extract.ts`).

**Comment by rarkins** [comment]
We can do without schedule now as there is no 22.04 for quite a while

**Review by viceice — APPROVED** [review]

File: lib/versioning/ubuntu/index.ts (hunk 3 above)
Comment (rarkins) [reply lib/versioning/ubuntu/index.ts:25]: Most version maps `isCompatible` to `isVersion`, which is what you're doing transitively

File: lib/versioning/ubuntu/index.ts [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+  range: string
+): string | null {
+  return versions.find((version) => equals(version, range)) ? range : null;
+}
+
+function minSatisfyingVersion(
+  versions: string[],
+  range: string
+): string | null {
+  return maxSatisfyingVersion(versions, range);
```
Comment (rarkins) [inline lib/versioning/ubuntu/index.ts:106]: maybe more intuitive to move this to a `findVersion()` function and have both max and min call that?

**Review by rarkins — CHANGES_REQUESTED** [review]

**Pull request merged**